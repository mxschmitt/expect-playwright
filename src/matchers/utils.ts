import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
