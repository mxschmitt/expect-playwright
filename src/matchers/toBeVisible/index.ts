import { SyncExpectationResult } from "expect/build/types"
import { getElement, getExpectedMessage, InputArguments } from "../utils"

const toBeVisible: jest.CustomMatcher = async function (
  ...args: InputArguments
): Promise<SyncExpectationResult> {
  try {
    const { elementHandle } = await getElement(...args)
    const actualVisibility = await elementHandle.isVisible()

    return {
      pass: actualVisibility,
      message: () =>
        getExpectedMessage(this, 'toBeVisible'),
    }
  } catch (err) {
    return {
      pass: false,
      message: () => err.toString(),
    }
  }
}

export default toBeVisible
