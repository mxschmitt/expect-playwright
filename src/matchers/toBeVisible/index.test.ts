import { assertSnapshot } from "../tests/utils"
import toBeVisible from "."

expect.extend({ toBeVisible })

describe("toBeVisible", () => {
  afterEach(async () => {
    await page.setContent("")
  })
  it("positive", async () => {
    await page.setContent(`<body><input id="foobar" style="visibility:visible"></body>`)
    await expect(page).toBeVisible("#foobar")
  })

  it("positive: handle-element", async () => {
    await page.setContent(`<input id="foobar" style="visibility:visible">`)
    await expect(await page.$('#foobar')).toBeVisible()
  })

  it("negative: target element is not visible", async () => {
    await page.setContent(`<input id="foo" style="visibility:hidden">`)
    await assertSnapshot(async () => expect(await page.$('#foo')).toBeVisible())
  })

  it("negative: handle-element : target element is not visible", async () => {
    await page.setContent(`<input id="foo" style="visibility:hidden">`)
    await assertSnapshot(async () => expect(await page.$('#foo')).toBeVisible())
  })

  describe("with 'not' usage", () => {
    it.only("positive", async () => {
      await page.setContent(`<body><div id="bar" style="visibility:hidden"><body>`)
      await expect(page).not.toBeVisible("div")
    })

    it("positive: handle-element", async () => {
      await page.setContent(`<input id="foo" style="visibility:hidden">`)
      await expect(await page.$('#foo')).not.toBeVisible()
    })

    it("negative: target element is visible", async () => {
      await page.setContent(`<input id="foobar" style="visibility:visible">`)
      await assertSnapshot(() => expect(page).not.toBeVisible("#foobar"))
    })

    it("negative: haandle-element: target element is visible", async () => {
      await page.setContent(`<input id="foobar" style="visibility:visible">`)
      await assertSnapshot(async () => expect(await page.$('#foobar')).not.toBeVisible())
    })
  })
  describe("timeout", () => {
    it("positive: should be able to use a custom timeout", async () => {
      setTimeout(async () => {
        await page.setContent(`<body><div id="foobar">Bar</div></body>`)
      }, 500)
      await expect(page).toBeVisible("#foobar")
    })
    it("should throw an error after the timeout exceeds", async () => {
      const start = new Date().getTime()
      await assertSnapshot(() =>
        expect(page).toBeVisible("#foobar", { timeout: 1 * 1000 })
      )
      const duration = new Date().getTime() - start
      expect(duration).toBeLessThan(1500)
    })
  })

})
