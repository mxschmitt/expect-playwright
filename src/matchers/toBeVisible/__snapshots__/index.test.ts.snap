// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeVisible negative: handle-element : target element is not visible 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoBeVisible[2m([22m[32mexpected[39m[2m)[22m

Expected: element to be visible"
`;

exports[`toBeVisible negative: target element is not visible 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoBeVisible[2m([22m[32mexpected[39m[2m)[22m

Expected: element to be visible"
`;

exports[`toBeVisible timeout should throw an error after the timeout exceeds 1`] = `"Error: Timeout exceed for element '#foobar'"`;

exports[`toBeVisible with 'not' usage negative: haandle-element: target element is visible 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoBeVisible[2m([22m[32mexpected[39m[2m)[22m

Expected: element to not be visible"
`;

exports[`toBeVisible with 'not' usage negative: target element is visible 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoBeVisible[2m([22m[32mexpected[39m[2m)[22m

Expected: element to not be visible"
`;
